### PR TITLE
Mask and unmasked forward mathematical equivalence

### PR DESCRIPTION
**Pull request by nlpfollower**
Run with:
`python train_dpo.py --model_name_or_path='meta-llama/Meta-Llama-3.1-8B-Instruct' --attn_implementation='flex_attention' --prefix_sharing --output_dir='output'`

For block mask visualization, `git clone https://github.com/pytorch-labs/attention-gym.git` into the parent directory, otherwise code will error.

And you get the following logs. Note the difference in the rejected portion of the context window.

```
Prompt + Chosen Comparison:
Overall - Max difference: 0.000000, Mean difference: 0.000000
Per-position statistics and next token predictions:
Position   0 - Max diff: 0.000000, Mean diff: 0.000000, Masked next: (id=14924), Unmasked next: (id=14924)
Position   1 - Max diff: 0.000000, Mean diff: 0.000000, Masked next: (id=374), Unmasked next: (id=374)
Position   2 - Max diff: 0.000000, Mean diff: 0.000000, Masked next: (id=264), Unmasked next: (id=264)
Position   3 - Max diff: 0.000000, Mean diff: 0.000000, Masked next: (id=6002), Unmasked next: (id=6002)
Position   4 - Max diff: 0.000000, Mean diff: 0.000000, Masked next: (id=4526), Unmasked next: (id=4526)
Position   5 - Max diff: 0.000000, Mean diff: 0.000000, Masked next: (id=358), Unmasked next: (id=358)
Position   6 - Max diff: 0.000000, Mean diff: 0.000000, Masked next: (id=13), Unmasked next: (id=13)
Position   7 - Max diff: 0.000000, Mean diff: 0.000000, Masked next: (id=1472), Unmasked next: (id=1472)
Position   8 - Max diff: 0.000000, Mean diff: 0.000000, Masked next: (id=279), Unmasked next: (id=279)
Position   9 - Max diff: 0.000000, Mean diff: 0.000000, Masked next: (id=3492), Unmasked next: (id=3492)
Position  10 - Max diff: 0.000000, Mean diff: 0.000000, Masked next: (id=304), Unmasked next: (id=304)
Position  11 - Max diff: 0.000000, Mean diff: 0.000000, Masked next: (id=720), Unmasked next: (id=720)
Position  12 - Max diff: 0.000000, Mean diff: 0.000000, Masked next: (id=11), Unmasked next: (id=11)
Position  13 - Max diff: 0.000000, Mean diff: 0.000000, Masked next: (id=220), Unmasked next: (id=220)
Position  14 - Max diff: 0.000000, Mean diff: 0.000000, Masked next: (id=18), Unmasked next: (id=18)
Position  15 - Max diff: 0.000000, Mean diff: 0.000000, Masked next: (id=11), Unmasked next: (id=11)
Position  16 - Max diff: 0.000000, Mean diff: 0.000000, Masked next: (id=220), Unmasked next: (id=220)
Position  17 - Max diff: 0.000000, Mean diff: 0.000000, Masked next: (id=1419), Unmasked next: (id=1419)
Position  18 - Max diff: 0.000000, Mean diff: 0.000000, Masked next: (id=11), Unmasked next: (id=11)
Position  19 - Max diff: 0.000000, Mean diff: 0.000000, Masked next: (id=220), Unmasked next: (id=220)
Position  20 - Max diff: 0.000000, Mean diff: 0.000000, Masked next: (id=2618), Unmasked next: (id=2618)
Position  21 - Max diff: 0.000000, Mean diff: 0.000000, Masked next: (id=11), Unmasked next: (id=11)
Position  22 - Max diff: 0.000000, Mean diff: 0.000000, Masked next: (id=220), Unmasked next: (id=220)
Position  23 - Max diff: 0.000000, Mean diff: 0.000000, Masked next: (id=374), Unmasked next: (id=374)
Position  24 - Max diff: 0.000000, Mean diff: 0.000000, Masked next: (id=264), Unmasked next: (id=264)
Position  25 - Max diff: 0.000000, Mean diff: 0.000000, Masked next: (id=4288), Unmasked next: (id=4288)
Position  26 - Max diff: 0.000000, Mean diff: 0.000000, Masked next: (id=1396), Unmasked next: (id=1396)
Position  27 - Max diff: 0.000000, Mean diff: 0.000000, Masked next: (id=10137), Unmasked next: (id=10137)
Position  28 - Max diff: 0.000000, Mean diff: 0.000000, Masked next: (id=13), Unmasked next: (id=13)
Position  29 - Max diff: 0.000000, Mean diff: 0.000000, Masked next: (id=33810), Unmasked next: (id=33810)

Prompt + Rejected Comparison:
Overall - Max difference: 8.812500, Mean difference: 0.342529
Per-position statistics and next token predictions:
Position   0 - Max diff: 0.000000, Mean diff: 0.000000, Masked next: (id=14924), Unmasked next: (id=14924)
Position   1 - Max diff: 0.000000, Mean diff: 0.000000, Masked next: (id=374), Unmasked next: (id=374)
Position   2 - Max diff: 0.000000, Mean diff: 0.000000, Masked next: (id=264), Unmasked next: (id=264)
Position   3 - Max diff: 0.000000, Mean diff: 0.000000, Masked next: (id=6002), Unmasked next: (id=6002)
Position   4 - Max diff: 0.000000, Mean diff: 0.000000, Masked next: (id=4526), Unmasked next: (id=4526)
Position   5 - Max diff: 0.000000, Mean diff: 0.000000, Masked next: (id=358), Unmasked next: (id=358)
Position   6 - Max diff: 0.000000, Mean diff: 0.000000, Masked next: (id=13), Unmasked next: (id=13)
Position   7 - Max diff: 0.000000, Mean diff: 0.000000, Masked next: (id=1472), Unmasked next: (id=1472)
Position   8 - Max diff: 0.000000, Mean diff: 0.000000, Masked next: (id=279), Unmasked next: (id=279)
Position   9 - Max diff: 0.000000, Mean diff: 0.000000, Masked next: (id=3492), Unmasked next: (id=3492)
Position  10 - Max diff: 0.000000, Mean diff: 0.000000, Masked next: (id=304), Unmasked next: (id=304)
Position  11 - Max diff: 0.000000, Mean diff: 0.000000, Masked next: (id=720), Unmasked next: (id=720)
Position  12 - Max diff: 5.015625, Mean diff: 0.846680, Masked next: (id=198), Unmasked next: (id=11)
WARNING! Token prediction mismatch at position 12 in Prompt + Rejected!
Position  13 - Max diff: 6.539062, Mean diff: 1.697266, Masked next: (id=220), Unmasked next: (id=220)
Position  14 - Max diff: 4.171875, Mean diff: 0.512207, Masked next: (id=1644), Unmasked next: (id=1114)
WARNING! Token prediction mismatch at position 14 in Prompt + Rejected!
Position  15 - Max diff: 5.199219, Mean diff: 1.316406, Masked next: (id=18), Unmasked next: (id=11)
WARNING! Token prediction mismatch at position 15 in Prompt + Rejected!
Position  16 - Max diff: 4.429688, Mean diff: 0.592285, Masked next: (id=220), Unmasked next: (id=220)
Position  17 - Max diff: 4.164062, Mean diff: 0.427002, Masked next: (id=9335), Unmasked next: (id=1313)
WARNING! Token prediction mismatch at position 17 in Prompt + Rejected!
Position  18 - Max diff: 3.453125, Mean diff: 0.743652, Masked next: (id=11), Unmasked next: (id=11)
Position  19 - Max diff: 2.884766, Mean diff: 0.419189, Masked next: (id=220), Unmasked next: (id=220)
Position  20 - Max diff: 2.437500, Mean diff: 0.308105, Masked next: (id=1313), Unmasked next: (id=1313)
Position  21 - Max diff: 3.328125, Mean diff: 0.550781, Masked next: (id=11), Unmasked next: (id=11)
Position  22 - Max diff: 2.296875, Mean diff: 0.336914, Masked next: (id=220), Unmasked next: (id=220)
Position  23 - Max diff: 2.837891, Mean diff: 0.484863, Masked next: (id=1313), Unmasked next: (id=1313)
Position  24 - Max diff: 2.796875, Mean diff: 0.400879, Masked next: (id=11), Unmasked next: (id=11)
Position  25 - Max diff: 2.500000, Mean diff: 0.348145, Masked next: (id=220), Unmasked next: (id=3639)
WARNING! Token prediction mismatch at position 25 in Prompt + Rejected!
Position  26 - Max diff: 1.242188, Mean diff: 0.191895, Masked next: (id=374), Unmasked next: (id=374)
Position  27 - Max diff: 1.292969, Mean diff: 0.192993, Masked next: (id=264), Unmasked next: (id=264)
Position  28 - Max diff: 1.071289, Mean diff: 0.173340, Masked next: (id=4288), Unmasked next: (id=4288)
Position  29 - Max diff: 0.986328, Mean diff: 0.162720, Masked next: (id=10137), Unmasked next: (id=10137)
Position  30 - Max diff: 1.294922, Mean diff: 0.387939, Masked next: (id=10137), Unmasked next: (id=10137)
Position  31 - Max diff: 8.812500, Mean diff: 0.929199, Masked next: (id=13), Unmasked next: (id=13)
Position  32 - Max diff: 2.320312, Mean diff: 0.279297, Masked next: (id=33810), Unmasked next: (id=33810)

Prompt length: 12
Chosen length: 18
Rejected length: 21
DPO sequence length: 128
Causal chosen sequence length: 128
Causal rejected sequence length: 128
Chosen index: 12
Rejected index: 30
```

### CODE REVIEW

**Comment by nlpfollower** [comment]
I wasn't sure how to clear the attention for the padding, but in my other repo I used a different mask_mod that removes the padding part, and the logits behaved similarly.

dpo_mask
causal_chosen_mask
causal_rejected_mask
